### ISSUE TITLE
FP S3937: Hexadecimal patterns are not handled properly

### ISSUE THREAD

**Issue by Corniel** [SonarSource/sonar-dotnet#3165]
### Description
For hexadecimal defined numbers (int and long), groups of 3 are okay, but 2 and 4 are not. I think that is not what should be the case.

### Expected behavior
``` C#
public static class S3937
{
    public static readonly long Hexa2 = 0xFF_FF_FF_FF; // compliant
    public static readonly long Hexa3 = 0x121_1F3_456; // non-compliant, hexadecimals should have groups of 2 or 4 
    public static readonly long Hexa4 = 0xF_FFFF_FFFF; // compliant
}
```
### Already okay
The binary defined work like I expected.
``` C#
public static class S3937
{
    public static readonly long Bin2 = 0b1_10_11_10_11; // compliant
    public static readonly long Bin3 = 0b0_111_111_111; // non-compliant, binaries should have groups of 2 or 4 
    public static readonly long Bin4 = 0b111_1101_1011; // compliant
}
```
* SonarC# Version 8.3

**Comment by christophe-zurn-sonarsource** [comment]
Hello @Corniel,

Thanks for reporting this.
While I agree on the FP part, I don't think the 3-digit groups for hexadecimals is a FN. According to the rule, it should raise on irregular number patterns. Although 3-digit groups for Hex and Binaries is weird IMO, it's still easier to read than irregular patterns and strictly valid when looking at the rule description.
Note that the `public static readonly long Bin3 = 0b0_111_111_111;` example you mentioned above does not currently raise any issue.

Here's in my opinion the current incorrect behaviour of the rules:
```
        public void S3937()
        {
            var hex1 = 0xFF_FF_FF_FA; // Compliant - this works as of now
            var hex2 = 0xFF_FF_FF_FD; // Noncompliant - FP
            var hex3 = 0xFF_FF_FF_FF; // Noncompliant - FP
            var hex4 = 0xFF_FF_FF_FFF; // Compliant- FN
        }
```

Current FP and FN come from the fact that the rule does not differentiate between decimal, hexadecimal and binary literals. The rule treats the last `D` and `F` found in the hexadecimal literals as a number suffix, which is obviously wrong.

**Comment by Corniel** [comment]
@christophe-zurn-sonarsource I would advise to not allow blocks bigger than 4 anyhow, and the first block should be allowed to be shorter. And groups of 3 for none-decimal patterns are I think so uncommon, that people should not use this rule or supress the ocation if the want that. But I the last thing is a personal view on what rules should bring, if you think otherwise, than we just agree to disagree on this one. ;)

**Comment by Corniel** [comment]
@christophe-zurn-sonarsource  Could you have a look to my pull request?

**Comment by andrei-epure-sonarsource** [comment]
> 
> 
> @christophe-zurn-sonarsource Could you have a look to my pull request?

there's no need to ping people, @Corniel . we monitor our open PRs.
